sigs/jvm/rfcs/20200126-tensorflow-data.md

```diff
@@ -0,0 +1,209 @@
+to the type of each component:
+
+Note: Each of the input components must share the same first "batch" dimension.
+
+```java
+Ops tf = // ... TensorFlow Ops accessor (either graph or eager)
+Dataset dataset = Dataset.fromTensorSlices(
+    Arrays.asList(tf.constant(features), tf.constant(labels)),
+    Arrays.asList(TInt32.DTYPE, TInt32.DTYPE)
+);
+```
+
+
+Other data sources are also possible, using `tf.data` ops. Datasets can also be constructed from 
+
+#### Transformations on Existing Datasets
+
+Once a dataset has been created from a data source it can be transformed by calling
+methods on the `Dataset` object. For example, to group elements in the above dataset into batches of size `2`, use `Dataset.batch(int batchSize)`:
+
+```java
+dataset = dataset.batch(2)
+```
+
+Dataset transformations alter both the values and shapes of the original elements, and
+return a *new* `Dataset` object.
+
+In this case, the original dataset had 4 elements of shape `[features: (3,) labels: (1,)]`.
+Once the `.batch` transformation is applied, the new dataset has 2 elements (batches) of shape `[features: (2, 3), labels: (2, 1)]`.
+
+Similar transformations include `.skip`, `.take`, `.map`, `.filter`, etc.
+
+
+### Iterating over Dataset Elements
+
+The primary use of a dataset is for iteration over its elements.
+Each row (or batch) element is represented as a list of tensor components, with
+type `List<Output<?>>`. The tensor components of this elements can be accessed using `List.get(int index)`.
+
+It is recommended to use `Tensor.expect(DataType<?> dtype)` to restore types
+to the retrieved tensors.
+
+
+#### Eager Mode: Iterable
+The `Dataset` class implements the `Iterable` interface, so in
+eager mode, iteration over dataset elements is possible using a standard for-each loop.
+
+Using the same example dataset from above, dataset elements can be extracted and
+used as follows:
+```java
+// Use default EagerSession
+Ops tf = Ops.create()
+
+// Dataset of (features, labels) from above
+Dataset dataset = Dataset.fromTensorSlices(tf, ... );
+
+// batch dataset elements into batches of size 2
+dataset = dataset.batch(2); 
+
+for (List<Output<?>> batch : dataset) {
+    Tensor<TInt32> featureBatch = element.get(0).tensor().expect(TInt32.DTYPE);
+    Tensor<TInt32> labelBatch = element.get(1).tensor().expect(TInt32.DTYPE);
+
+    // Perform batch-wise computations on featreBatch and labelBatch
+}
+
+```
+
+
+#### Graph Mode: OneShotIterator
+
+The above code will not work in graph mode, which requires the use of `Session`s
+to run the computations. In graph mode, datasets can be iterated over using the `OneShotIterator` abstraction, and a while loop.
+
+`OneShotIterator` allows iteration over a dataset's elements "one time", using
+repeated calls to `Session.run`. 
+
+The `OneShotIterator` object has two fields:
+- `OneShotIterator.getMakeIteratorOp()` returns a "make-iterator" `Operation` which is run to initialize this iterator
```

Comment:
might be interesting to see where that will fit with changes from Adam, he add something related to graph initialization run.

Reply:
Yup, I added a `DatasetIterator` to replace the `OneShotIterator` class (more closely mimics the `tf.data.Iterator` from Python too). It uses the new `tf.initAdd` / `tf.init`!